File: SeiveOfEratosthenes.py
'''
@author Matthew Pecko
An algorithm to perform a seive of Eratosthenes. Each number is represented as a Number object.
'''
import math

# Simple data structure to store the representaion and value of a number
class MyNumber:
    number = None
    value = None
    def __init__(self, number, value):
        self.number = number
        self.value = value

    def __str__(self):
        return "{} \t {} ".format(self.number, self.value)

    def getNumber(self):
        return self.number

    def getValue(self):
        return self.value

    def setValue(self, value):
        self.value = value

    def printFormatted(self):
        print("My number is: ", self.number, " | My value is: ", self.value)



def main():
    '''
    Overhead loop
    Start with the first prime and go to MAX + 1 to make MAX inclusive, storing a True value for every number.
    '''
    MIN = 2
    MAX = 2**16

    A = []

    for n in range(MIN, MAX+1):
        A.append(MyNumber(n, True))


    # Do not exceed the sqrt of the size of list because ??? wikipedia pseudocode told me so..
    size = int(math.sqrt(len(A)))

    i = 0
    while A[i].getNumber() <= size:
        if A[i].getValue() is True:
            # Start at the prime number squared for optimization, and keep track of a counter to include all multiples of the prime
            j = A[i].getNumber()**2
            counter = 0

            # go to length of the list plus 1 to insure the last number is tested and not skipped
            while j <= len(A) + 1:

                # Becuase the index of the list starts with 0 and not 2, I have to move the pointer two indicies before j for alignment.
                A[j-2].setValue(False)

                # Increment j by a multiple of the prime
                counter += 1
                j = A[i].getNumber()**2 + counter * A[i].getNumber()
        i += 1

    for n in A:
        if n.getValue():
            print(n)

if __name__ == "__main__":
    main()

File: SeiveOfEratosthenes.c
/*
 *
 * @author Matthew Pecko
 * 12/29/2020
 *
 * Seive of Eratosthenes in C
 *
 * Define MIN as 2, and MAX to be the upper bound
 * of the prime range (in decimal)
 *
 *
 * The number of bytes we will need can be found:
 * numberOfBytes = diff/32
 * trueNumber = floor(numberOfBytes)+1
 *
 *
 * We want to seive the range from [0,sqrt(diff)]
 * we will stick with [0,diff] for now
 *
 */


#include "stdlib.h"
#include "stdio.h"
#include "stdint.h"
#include "math.h"

#define MIN 2
#define MAX 999999999

void print(uint32_t* a);

int main(){
	uint32_t size, i, j, counter, p, k, e;
	size = (uint32_t)floor(MAX/32);
	uint32_t* state = calloc(size+1, 4);
	i = 2;
	while(i<=MAX){
		k = i;
		p = k % 32;
		e = (uint32_t)(k / 32);
		if(!(state[e] & (1 << p))){
			j = i*i;
			counter = 0;
			while(j <= MAX+1){
				k = j;
				p = k % 32;
				e = (uint32_t)(k / 32);
				state[e] = state[e] | (1 << (p));
				counter++;
				j = (i*i) + counter*i;
			}	
		}
		i++;
	}
	printf("\n\n\n\n\n\n");
	printf("RESULT--------------------------- \n");
	print(state);
}

void print(uint32_t* a){
	uint32_t i, k, p, e;
	i = 2;
	do{
		k = i;
		p = k % 32;
		e = (uint32_t)(k / 32);
		if(!(a[e] & (1 << p))){
			printf("The state of %d is: \t true \n", i);
		}else{
			printf("The state of %d is: \t false \n", i);
		}	
		i++;
	}while(i <= MAX);
}
